File: frontend/src/lib/platforms.ts
import catalog from "./platformCatalog.json" with { type: "json" }
import type { Platform, PlatformCapability, PlatformId } from "@/types"

/**
 * 平台接入能力的唯一配置源。界面只能按这里声明的能力开放入口，
 * 避免新增平台时仅补一个图标便被误认为已经接入。
 */
export const PLATFORMS: readonly Platform[] = catalog as readonly Platform[]

/** 使用原生来源的平台；Gemini / Grok 另有官方 API Key 只读检测，须同时检查连接 kind */
export const FETCH_ONLY_PLATFORMS: readonly PlatformId[] = ["gemini", "grok", "zcode", "trae", "qoder", "workbuddy"]

export function isFetchOnlyPlatform(id: PlatformId): boolean {
  return FETCH_ONLY_PLATFORMS.includes(id)
}

export function platformConfig(id: PlatformId): Platform {
  return PLATFORMS.find((platform) => platform.id === id) ?? PLATFORMS[0]
}

export function platformSupports(id: PlatformId, capability: PlatformCapability): boolean {
  return platformConfig(id).capabilities.includes(capability)
}

export function platformAvailabilityText(platform: Platform): string {
  if (platform.availability === "available") return "可用"
  if (platform.availability === "not-connected") return "未连接"
  return "待接入"
}

File: backend/src/platforms.rs
//! 前后端共用的平台目录。来源是否存在由本机检测决定，不能由图标推断。
use serde::Deserialize;
use std::sync::LazyLock;

#[derive(Deserialize)]
pub struct Platform {
    pub id: String,
    pub name: String,
    pub capabilities: Vec<String>,
    pub limitation: Option<String>,
}

static CATALOG: LazyLock<Vec<Platform>> = LazyLock::new(|| {
    serde_json::from_str(include_str!("../../frontend/src/lib/platformCatalog.json"))
        .expect("随程序打包的平台目录必须有效")
});

pub fn get(id: &str) -> Option<&'static Platform> {
    CATALOG.iter().find(|p| p.id == id)
}
pub fn known(id: &str) -> bool {
    get(id).is_some()
}
pub fn native(id: &str) -> bool {
    known(id) && !matches!(id, "claude" | "codex")
}
pub fn supports(id: &str, capability: &str) -> bool {
    get(id).is_some_and(|p| p.capabilities.iter().any(|c| c == capability))
}
pub fn limitation(id: &str) -> String {
    get(id)
        .and_then(|p| p.limitation.clone())
        .unwrap_or_else(|| "该平台未提供此项数据".into())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn all_eight_platforms_have_unique_ids() {
        assert_eq!(CATALOG.len(), 8);
        for id in [
            "claude",
            "codex",
            "gemini",
            "grok",
            "zcode",
            "trae",
            "qoder",
            "workbuddy",
        ] {
            assert_eq!(CATALOG.iter().filter(|p| p.id == id).count(), 1);
        }
        assert!(!known("tare"));
        assert!(!supports("trae", "local_sessions"));
    }
}
